File: C++/Linked-List/Detect Loop in Linked List.cpp
// C++ program to detect loop in a linked list
#include <bits/stdc++.h>
using namespace std;

/* Link list node */
class Node {
public:
	int data;
	Node* next;

    /* constructor */
    Node(int value) {
        data = value;
        next = NULL;
    }
};

int detectLoop(Node* list)
{
    /* create two pointers slow and fast
    slow will move by 1 step while fast will move by 2 steps. */
	Node *slow = list, *fast = list;

	while (fast && fast -> next) {
		slow = slow -> next;
		fast = fast -> next -> next;

        /* at any point slow is equal to fast that means loop is present so return true */
		if (slow == fast) {
			return 1;
		}
	}

	return 0;
}

int main()
{
	/* Start with the empty list */
	Node* head = NULL;
    Node* tail = head;

	while(true) {
        int data;
        cin >> data;    /* taling inpt of data from user */

        /* assuming -1 as NULL/end of list */
        if(data == -1) {
            break;
        }
        else {
            /* create first node */
            Node* new_node = new Node(data);

            if(head == NULL) {
                /* point head to first node */
                head = tail = new_node;
            }
            else {
                /* link the old list of the new node */
                tail -> next = new_node;
                
                /* move tail to the new node */
                tail = tail -> next;
            }
        }
    }

	/* Create a loop for testing */
	tail -> next = head;

	if (detectLoop(head)) {
		cout << "Loop Found";
    }
	else {
		cout << "No Loop found";
    }
	return 0;
}

// Time complexity: O(N)
// Auxiliary Space: O(1)

File: Python/Linked-List/Detect Loop in Linked list.py
"""
Python program to detect a loop in a linked list.
A loop is detected if there is a cycle in the linked list.

Algorithm:
1. Create two pointers, slow and fast. Both initially point to the head of the linked list.
2. Move slow pointer by one step and fast pointer by two steps.
3. If there is a loop, slow and fast pointers will meet at some point.
   This is because if there is a cycle, the fast pointer will eventually catch up with the slow pointer.
4. If slow and fast pointers meet, it means there is a loop in the linked list.
5. If fast pointer reaches the end of the list (NULL), it means there is no loop.

Time complexity: O(N)
Space complexity: O(1)
"""

class Node:
    def __init__(self, value):
        self.data = value
        self.next = None

def detectLoop(head):
    # Create two pointers, slow and fast
    slow = head
    fast = head

    while fast and fast.next:
        # Move slow pointer by one step
        slow = slow.next
        # Move fast pointer by two steps
        fast = fast.next.next

        # If slow and fast pointers meet, there is a loop
        if slow == fast:
            return True

    # No loop found
    return False

# Create linked list
head = None
tail = None

while True:
    data = int(input("Enter data (enter -1 to end): "))

    if data == -1:
        break
    else:
        # Create a new node
        new_node = Node(data)

        if head is None:
            # First node in the list
            head = tail = new_node
        else:
            # Append new node to the list
            tail.next = new_node
            tail = new_node

# Create a loop for testing
tail.next = head

if detectLoop(head):
    print("Loop found")
else:
    print("No loop found")
